server/src/main/java/org/elasticsearch/ingest/IngestStats.java

```diff
@@ -263,20 +262,65 @@ static List<PipelineStat> merge(List<PipelineStat> first, List<PipelineStat> sec
      */
     public record ProcessorStat(String name, String type, Stats stats) {
 
-        // The list of ProcessorStats has *always* stats for each processor (even if processor was executed or not), so it's safe to zip
-        // both lists using a common index iterator.
         private static List<ProcessorStat> merge(List<ProcessorStat> first, List<ProcessorStat> second) {
-            var merged = new ArrayList<ProcessorStat>();
-            assert first.size() == second.size()
-                : "stats size mismatch ["
-                    + first.stream().map(ps -> ps.name + ":" + ps.type).collect(Collectors.joining(","))
-                    + "] ["
-                    + second.stream().map(ps -> ps.name + ":" + ps.type).collect(Collectors.joining(","))
-                    + "]";
-            for (var i = 0; i < first.size(); i++) {
-                merged.add(new ProcessorStat(first.get(i).name, first.get(i).type, Stats.merge(first.get(i).stats, second.get(i).stats)));
+            // in the simple case, this amounts to summing up the stats in the first and second and returning
+            // a new list of stats that contains the sum. but there are a few not-quite-so-simple cases, too,
+            // so this logic is a little bit intricate.
+
+            // total up the stats across both sides
+            long firstIngestCountTotal = 0;
+            for (ProcessorStat ps : first) {
+                firstIngestCountTotal += ps.stats.ingestCount;
+            }
+
+            long secondIngestCountTotal = 0;
+            for (ProcessorStat ps : second) {
+                secondIngestCountTotal += ps.stats.ingestCount;
+            }
+
+            // early return in the case of a non-ingest node (the sum of the stats will be zero, so just return the other)
+            if (firstIngestCountTotal == 0) {
+                return second;
+            } else if (secondIngestCountTotal == 0) {
+                return first;
+            }
+
+            // the list of stats can be different depending on the exact order of application of the cluster states
+            // that apply a change to a pipeline -- figure out if they match or not (usually they match!!!)
+            boolean match = true;
+            if (first.size() != second.size()) {
+                match = false;
+            } else {
+                for (var i = 0; i < first.size(); i++) {
```

Comment:
This code is not really performance sensitive, is it? I don't think it is. But if it is, we could speculatively compute the merged result in this loop, and then bail out in the rare case that types don't match.